pkg/zkproofs/range.go

```diff
@@ -0,0 +1,144 @@
+package zkproofs
+
+import (
+	crand "crypto/rand"
+	"encoding/json"
+	"errors"
+	"github.com/coinbase/kryptology/pkg/bulletproof"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/gtank/merlin"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption/elgamal"
+)
+
+const rangeDomain = "rangeDomain"
+const ippDomain = "ippDomain"
+
+type RangeProof struct {
+	Proof      *bulletproof.RangeProof
+	Randomness curves.Point
+	UpperBound int
+}
+
+// NewRangeProof a range proof for some ciphertext that proves the value is between 0 and 2^upperBound
+// Note that while this proof generator takes in the plaintext of the value, it will only work in conjunction with a
+// ciphertext that encrypts value using the given randomness.
+func NewRangeProof(upperBound, value int, randomness curves.Scalar) (*RangeProof, error) {
+	if randomness == nil {
+		return nil, errors.New("invalid randomness factor")
+	}
+	curve := curves.ED25519()
+	prover, err := bulletproof.NewRangeProver(upperBound, getRangeDomain(), getIppDomain(), *curve)
+	if err != nil {
+		return nil, err
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+	g := eg.GetG()
+	h := eg.GetH()
+	u := curve.Point.Random(crand.Reader)
+	proofGenerators := bulletproof.NewRangeProofGenerators(g, h, u)
+	transcript := getTranscript()
+
+	vScalar := curve.Scalar.New(value)
+	proof, err := prover.Prove(vScalar, randomness, upperBound, proofGenerators, transcript)
+	if err != nil {
+		return nil, err
+	}
+	return &RangeProof{
+		Proof:      proof,
+		Randomness: u,
+		UpperBound: upperBound,
+	}, nil
+}
+
+// VerifyRangeProof verifies the range proof for the given ciphertext
+func VerifyRangeProof(proof *RangeProof, ciphertext *elgamal.Ciphertext) (bool, error) {
```

Comment:
Similar comment as above, we could consider adding the upperBound to this method as well, since it should be up to the verifier to determine what upper bound they want to check for (just that the associated proof should not work if it's not meant for that upper bound)

Reply:
Added